### PR TITLE
Add knowledge of branch freezes to the Release model

### LINKED ISSUE #105: don't allow updates to be stuck "between" repos while in "freeze"
Currently it is possible while in freeze to have update/s which are "stuck" between repos. 
- Maintainer submits an update.
- Update receives enough karma to get to stable before it get pushed to 'updates-testing' 
- Update don't appear in any repo untill freeze is over and it gets pushed to stable repos ('fedora'/'updates')

I think this will improve security (all updates are available in freeze) without manually downloading them from 'koji', also this helps with testing because not everyone downloads packages manually and more people testing things is better.

### LINKED ISSUE #1038: Bodhi should inform about freeze
[Miro Hrončok reported this issue in Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1026321)

> Description of problem:
> When there is any freeze that prevents the update to be pushed in stable, bodhi should inform about > that.
> 
> Steps to Reproduce:
> 1. While there is a freeze, mark an update as stable
> 
> Actual results:
> Bodhi says:
> This update has been submitted for stable by $USERNAME. 
> 
> But the update is not pushed, because of the freeze. When the user doesn't know about the freeze,
> (s)he is confused.
> 
> Expected results:
> Bodhi says:
> This update has been submitted for stable by $USERNAME.
> There is an ongoing freeze, this will be pushed to stable after the freeze is over. $LINK_TO_WIKI

### LINKED ISSUE #1563: Add knowledge of branch freezes to the Release model
We should add the ability for Bodhi to know when a branched Release is frozen. During freeze periods, the masher should avoid mashing updates that are request stable, but it should allow exceptions by looking for blocker bugs. This way, releng doesn't have to run ```bodhi-push``` two times (once for the f<branched>-updates-testing, and again for all other stable releases).

### PR DESCRIPTION
**Pull request by sebwoj**
fix #105
fix #1038
fix #1563

Signed-off-by: Sebastian Wojciechowski <s.wojciechowski89@gmail.com>

### CODE REVIEW

File: bodhi/server/services/releases.py [hunk 1]
```diff
@@ -338,6 +339,22 @@ def save_release(request):
...
+                        Update.request == UpdateRequest.stable
+                    ).filter(
+                        Update.locked == False
+                    ).all()
+                    for u in updates:
+                        u.comment(
+                            request.db,
+                            'There is an ongoing freeze, this will be pushed to'
+                            ' stable after the freeze is over.',
+                            author='bodhi',
```
Comment (bowlofeggs) [inline bodhi/server/services/releases.py:356]: Nice touch!

File: bodhi/server/services/releases.py [hunk 2]
```diff
@@ -338,6 +339,22 @@ def save_release(request):
...
+                        r.state != ReleaseState.frozen:
+                    updates = request.db.query(Update).filter(Update.release_id == r.id).filter(
+                        Update.request == UpdateRequest.stable
+                    ).filter(
+                        Update.locked == False
+                    ).all()
+                    for u in updates:
+                        u.comment(
+                            request.db,
+                            'There is an ongoing freeze, this will be pushed to'
```
Comment (bowlofeggs) [inline bodhi/server/services/releases.py]:
```suggestion
                            'There is an ongoing freeze; this will be pushed to'
```

**Review by bowlofeggs — COMMENTED** [review]
@mohanboddu - do you think it would be helpful if ```bodhi-push``` automatically filtered out stable updates for frozen releases when the --builds flag is not used?

**Comment by sebwoj** [comment]
1) I forgot to mention that there are some places where `models.ReleaseState.frozen` could be(or not) inserted, but I'm not sure. Example:
https://github.com/fedora-infra/bodhi/blob/0c382987cc7ba1a08c58d227bb0388472773792d/bodhi/server/scripts/untag_branched.py#L77

2) > @mohanboddu - do you think it would be helpful if bodhi-push automatically filtered out stable updates for frozen releases when the --builds flag is not used?

Yeah, good question.

**Comment by mohanboddu** [comment]
> @mohanboddu - do you think it would be helpful if `bodhi-push` automatically filtered out stable updates for frozen releases when the --builds flag is not used?

@bowlofeggs I dont exactly understand your question, we use --builds flag only to push certain builds to stable (not all eligible stable updates) for the frozen release.

We currently use --request flag to push only testing updates for the frozen release.

**Comment by bowlofeggs** [comment]
@mohanboddu The idea is that you could just run bodhi-push the same during a freeze as you do during other times, and since Bodhi will know about freezes it will avoid pushing updates to stable by default.

**Comment by sebwoj** [comment]
@bowlofeggs What do you think about @nirik idea to check state(daily script) using the following url:
https://infrastructure.fedoraproject.org/infra/ansible/vars/all/Frozen.yaml

We can use also:
https://infrastructure.fedoraproject.org/infra/ansible/vars/all/00-FedoraCycleNumber.yaml

**Comment by bowlofeggs** [comment]
@sebwoj I'm unsure about that. The problem is that it ties upstream Bodhi into a particular way that Infrastructure does Ansible, which won't make sense for other deployments (believe it or not, there are other Bodhi deployments than Fedora's).

It's better to let the Ansible playbook use whatever we do here than it is for us to try to interpret the Ansible playbook's yaml files.

**Review by bowlofeggs — COMMENTED** [review]
I'm happy to accept this as is. @sebwoj are you ready for this to get merged?

File: bodhi/server/services/releases.py (hunk 1 above)
Comment (sebwoj) [reply bodhi/server/services/releases.py:356]: Thanks :)

**Comment by sebwoj** [comment]
PR rebased.
I'm ok with current version, bur we should be sure about

> 1. I forgot to mention that there are some places where `models.ReleaseState.frozen` could be(or not) inserted, but I'm not sure. Example:
>    https://github.com/fedora-infra/bodhi/blob/0c382987cc7ba1a08c58d227bb0388472773792d/bodhi/server/scripts/untag_branched.py#L77

**Comment by bowlofeggs** [comment]
@sebwoj to this day, I'm not sure what the untag_branched script is really for. I'm not sure whether it is used or not. So… I think we can ignore that for now.

**Review by bowlofeggs — APPROVED** [review]
Thanks!

**Pull request merged**